### PR TITLE
RFI reader and miscellaneous metadata reader for CARD-NRB compliance

### PR DESCRIPTION
**Pull request by seongsujeong**
This PR adds `s1-reader` an ability to load Radio Frequency Interferences (RFI) information, which is available starting from IPF version 3.40.

edit: 03/31/2023
This PR also includes the feature add to load miscellaneous metadata to populate the metadata of RTC-S1 to comply with [CEOS Analysis Ready Data](https://ceos.org/ard/) - Normalized Radar Backscatter (CARD-NRB)

### CODE REVIEW

**Comment by seongsujeong** [comment]
@gshiroma @vbrancat This PR will be used to extract RFI information to be populated in the RTC and CSLC product.
Would you mind taking a look at this PR? Thanks

File: src/s1reader/s1_annotation.py [hunk 1]
```diff
@@ -515,6 +576,156 @@ def load_from_zip_file(cls, path_aux_cal_zip: str, pol: str, str_swath: str):
 
         return cls
 
+
+@dataclass
+class SwathRfiInfo:
```
Comment (gshiroma) [inline src/s1reader/s1_annotation.py:581]:
```suggestion
class SwathRFIInfo:
```

File: src/s1reader/s1_annotation.py [hunk 2]
```diff
@@ -7,20 +7,75 @@
...
+    "azimuthTime": lambda T: datetime.datetime.strptime(T, '%Y-%m-%dT%H:%M:%S.%f'),
+    "inBandOutBandPowerRatio": float,
+    "percentageAffectedLines": float,
+    "avgPercentageAffectedSamples": float,
+    "maxPercentageAffectedSamples": float,
+    "numSubBlocks": int,
+    "subBlockSize": int,
+    "maxPercentageAffectedBW": float,
+    "percentageBlocksPersistentRfi": float,
+    "maxPercentageBWAffectedPersistentRfi": float
```
Comment (gshiroma) [inline src/s1reader/s1_annotation.py:41]:
```suggestion
    "maxPercentageBWAffectedPersistentRFI": float
```

**Review by gshiroma — COMMENTED** [review]
Thank you so much for this updates, @seongsujeong . We talked offline, but I'll leave some of the comments we talked about here for reference.

File: src/s1reader/s1_annotation.py [hunk 3]
```diff
@@ -7,20 +7,75 @@
...
 from packaging import version
+from scipy.interpolate import InterpolatedUnivariateSpline, interp1d
 
-from isce3.core import speed_of_light
 
 # Minimum IPF version from which the S1 product's Noise Annotation
 # Data Set (NADS) includes azimuth noise vector annotation
 min_ipf_version_az_noise_vector = version.parse('2.90')
 
+# Minimum IPF version from which the RFI information gets available
```
Comment (vbrancat) [inline src/s1reader/s1_annotation.py:26]: Can we add a link to a document where this is stated?

File: src/s1reader/s1_annotation.py [hunk 4]
```diff
@@ -7,20 +7,75 @@
...
+    "swath": str,
+    "azimuthTime": lambda T: datetime.datetime.strptime(T, '%Y-%m-%dT%H:%M:%S.%f'),
+    "inBandOutBandPowerRatio": float,
+    "percentageAffectedLines": float,
+    "avgPercentageAffectedSamples": float,
+    "maxPercentageAffectedSamples": float,
+    "numSubBlocks": int,
+    "subBlockSize": int,
+    "maxPercentageAffectedBW": float,
+    "percentageBlocksPersistentRfi": float,
```
Comment (vbrancat) [inline src/s1reader/s1_annotation.py:40]:
```suggestion
    "percentageBlocksPersistentRFI": float,
```
For consistency with Gustavo's comment.

File: src/s1reader/s1_reader.py [hunk 5]
```diff
@@ -289,6 +290,50 @@ def get_ipf_version(tree: ET):
 
     return ipf_version
 
+
+def get_swath_misc_metadata(et_manifest: ET, et_product:ET,
+                            product_annotation:ProductAnnotation):
+    '''
+    Extract the miscellaneous metadata to populate CARD4L-NRB metadata
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/s1reader/s1_reader.py:297]: Let's be more descriptive on what are this metadata (maybe provide a couple of examples)

**Review by vbrancat — COMMENTED** [review]
Thank you @seongsujeong I only have minor comments on the code. Hopefully I can find some time to test the code.

**Comment by vbrancat** [comment]
@seongsujeong, any progress on this PR?

File: src/s1reader/s1_annotation.py (hunk 4 above)
Comment (seongsujeong) [reply src/s1reader/s1_annotation.py:40]: The keys in `dict_datatype_rfi` have to be in exact match with the field names in RFI annotation .xml files because the following function `element_to_dict()` make use of the dictionary to parse the XML.

File: src/s1reader/s1_annotation.py (hunk 2 above)
Comment (seongsujeong) [reply src/s1reader/s1_annotation.py:41]: Same comment as above: The keys in `dict_datatype_rfi` have to be in exact match with the field names in RFI annotation .xml files because the following function `element_to_dict()` make use of the dictionary to parse the XML.

File: src/s1reader/s1_annotation.py (hunk 3 above)
Comment (seongsujeong) [reply src/s1reader/s1_annotation.py:26]: Good idea. I've added few information about the document I've referenced.

File: src/s1reader/s1_reader.py [hunk 6]
```diff
@@ -523,11 +567,17 @@ def burst_from_xml(annotation_path: str, orbit_path: str, tiff_path: str,
                                             'annotation/rfi/rfi-')
             with open_method(rfi_annotation_path, 'r') as f_rads:
                 tree_rads = ET.parse(f_rads)
-                burst_rfi_info_swath = SwathRfiInfo.from_et(tree_rads, tree_lads, ipf_version)
+                burst_rfi_info_swath = SwathRfiInfo.from_et(tree_rads,
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/s1reader/s1_reader.py]: Done

**Comment by seongsujeong** [comment]
@vbrancat @gshiroma Would you mind taking another look at the PR? Thanks!

File: src/s1reader/s1_annotation.py [hunk 7]
```diff
@@ -7,20 +7,78 @@
...
+    '''
+    if dict_tree is None:
+        dict_tree = {}
+    key_elem = elem_in.tag
+    child_elem = list(elem_in.iterchildren())
+
+    if len(child_elem) == 0:
+        # Reached the tree end
+        text_elem = elem_in.text
+        dict_tree[key_elem] = dict_datatype_rfi[key_elem](text_elem)
```
Comment (gshiroma) [inline src/s1reader/s1_annotation.py:73]: If they update their metadata with a new field, I think this line will break. You probably want to test if `key_elem` is a member of `dict_datatype_rfi.keys()`. If not, you probably want to print a warning message so we know we may need to add a new element to `dict_datatype_rfi`.

File: src/s1reader/s1_annotation.py [hunk 8]
```diff
@@ -515,6 +579,167 @@ def load_from_zip_file(cls, path_aux_cal_zip: str, pol: str, str_swath: str):
 
         return cls
 
+
+@dataclass
+class SwathRFIInfo:
```
Comment (gshiroma) [inline src/s1reader/s1_annotation.py:584]: Sorry for the "back and fourth" regarding these acronyms camel case convention, but since we decided to use "Rfi" in `dict_datatype_rfi`, I think we should also revert the changes to the class name here from `SwathRFIInfo` to `SwathRfiInfo`.

File: src/s1reader/s1_annotation.py (hunk 2 above)
Comment (gshiroma) [reply src/s1reader/s1_annotation.py:44]: Sounds good. Thank you, @seongsujeong !

File: src/s1reader/s1_reader.py [hunk 9]
```diff
@@ -17,9 +17,10 @@
 
 from s1reader import s1_annotation  # to access __file__
 from s1reader.s1_annotation import ProductAnnotation, NoiseAnnotation,\
-                                   CalibrationAnnotation, AuxCal, \
+                                   CalibrationAnnotation, AuxCal,\
                                    BurstCalibration, BurstEAP, BurstNoise,\
-                                   BurstExtendedCoeffs
+                                   BurstExtendedCoeffs, SwathRFIInfo, SwathMiscMetadata,\
```
Comment (gshiroma) [inline src/s1reader/s1_reader.py:22]:
```suggestion
                                   BurstExtendedCoeffs, SwathRfiInfo, SwathMiscMetadata,\
```

File: src/s1reader/s1_reader.py [hunk 10]
```diff
@@ -502,19 +555,37 @@ def burst_from_xml(annotation_path: str, orbit_path: str, tiff_path: str,
...
         product_annotation = ProductAnnotation.from_et(tree_lads)
 
+        # Load RFI information if available
+        if ipf_version >= RFI_INFO_AVAILABLE_FROM:
+            rfi_annotation_path =\
+                    annotation_path.replace('annotation/',
+                                            'annotation/rfi/rfi-')
+            with open_method(rfi_annotation_path, 'r') as f_rads:
+                tree_rads = ET.parse(f_rads)
+                burst_rfi_info_swath = SwathRFIInfo.from_et(tree_rads,
```
Comment (gshiroma) [inline src/s1reader/s1_reader.py:577]:
```suggestion
                burst_rfi_info_swath = SwathRfiInfo.from_et(tree_rads,
```

**Review by gshiroma — COMMENTED** [review]
Looks great, @seongsujeong . I have just a few minor comments below. I think we are close to merge it. Thanks!

File: src/s1reader/s1_annotation.py (hunk 8 above)
Comment (seongsujeong) [reply src/s1reader/s1_annotation.py:584]: Makes sense. We should stick to the one convention.

File: src/s1reader/s1_annotation.py (hunk 7 above)
Comment (seongsujeong) [reply src/s1reader/s1_annotation.py]: Good idea. I've added the warning message in the case above. Also, the values will be treated as `str` when there is no castor defined in `dict_datatype_rfi`

File: src/s1reader/s1_annotation.py [hunk 11]
```diff
@@ -7,20 +7,86 @@
...
+    if len(child_elem) == 0:
+        # Reached the tree end
+        text_elem = elem_in.text
+
+        if key_elem in dict_datatype_rfi:
+            value_castor = dict_datatype_rfi[key_elem]
+        else:
+            warnings.warn(f'Datetype Castor for {key_elem} was not defined. Considering the value as string.')
+            value_castor = str
+        dict_tree[key_elem] = value_castor(text_elem)
```
Comment (gshiroma) [inline src/s1reader/s1_annotation.py:80]:
```suggestion
        if key_elem in dict_datatype_rfi:
            elem_datatype = dict_datatype_rfi[key_elem]
        else:
            warnings.warn(f'Data type for element {key_elem} is not defined. Casting the value "{text_elem}" as string.')
            elem_datatype = str
        dict_tree[key_elem] = elem_datatype(text_elem)
```

File: src/s1reader/s1_annotation.py [hunk 12]
```diff
@@ -73,11 +73,11 @@ def element_to_dict(elem_in: ET, dict_tree: dict = None):
         text_elem = elem_in.text
 
         if key_elem in dict_datatype_rfi:
-            value_castor = dict_datatype_rfi[key_elem]
+            elem_datatype = dict_datatype_rfi[key_elem]
         else:
-            warnings.warn(f'Datetype Castor for {key_elem} was not defined. Considering the value as string.')
-            value_castor = str
-        dict_tree[key_elem] = value_castor(text_elem)
+            warnings.warn(f'Data type for element {key_elem} is not defined. Casting the value "{text_elem}" as string.')
```
Comment (gshiroma) [inline src/s1reader/s1_annotation.py:78]: Thank you, @seongsujeong . Could you also split this line so it's not too long?

Comment (seongsujeong) [reply src/s1reader/s1_annotation.py]: Absolutely. I've split the warning message into two lines.

**Review by gshiroma — APPROVED** [review]

**Review by vbrancat — APPROVED** [review]
LGTM :)
